core/src/main/java/com/confighub/core/utils/PropertiesSubstitutor.java

```diff
@@ -0,0 +1,101 @@
+package com.confighub.core.utils;
+
+import com.confighub.core.repository.Property;
+import com.confighub.core.repository.PropertyKey;
+import com.confighub.core.security.SecurityProfile;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.*;
+import java.util.regex.Pattern;
+
+public class PropertiesSubstitutor {
+
+    private static final Logger log = LogManager.getLogger(PropertiesSubstitutor.class);
+
+    private static String replaceKey(String key, String value, String text)
+    {
+        return text.replaceAll("\\$\\{" + Pattern.quote(key) + "}", value);
+    }
+
+    private static boolean containsKey(String key, String text)
+    {
+        return text.contains("${" + key + "}");
+    }
+
+    public static Map<PropertyKey, String> resolveTextSubstitutions(final Map<PropertyKey, Property> resolved,
```

Comment:
I am only concerned with the amount of extra work this is doing to handle things that refer to eachother.  We've been dealing with performance issues and would like to potentially limit the amount of work necessary to resolve a file.  This seems like it is adding a significant amount of work to solve a case that can be solved by not having configurations bidirectional.  I'm curious on Edin's point of view here.